runtime/new-recipe.js

```diff
@@ -15,30 +15,106 @@ class Connection extends Edge {
 }
 
 class Particle extends Node {
-  get id() {} // Not resolved until we have an ID.
-  get name() {}
-  get tags() {}
-  get providedSlots() {} // Slot*
-  get consumedSlots() {} // SlotConnection*
-  get parameters() {} // {parameter -> ViewConnection}
+  constructor(recipe, name) {
+    assert(recipe);
+    this._recipe = recipe;
+    this._id = undefined;
+    this._name = name;
+    this._tags = [];
+    this._providedSlots = [];
+    this._consumedSlots = [];
+    this._connections = {};
+    this._unnamedConnections = [];
+  }
+  get id() { return this._id; } // Not resolved until we have an ID.
+  get name() { return this._name; }
+  get tags() { return this._tags; }
+  get providedSlots() { return this._providedSlots; } // Slot*
+  get consumedSlots() { return this._consumedSlots; } // SlotConnection*
+  get connections() { return this._connections; } // {parameter -> ViewConnection}
+
+  addUnnamedConnection() {
+    var connection = new ViewConnection(undefined, this);
+    this._unnamedConnections.push(connection);
+    return connection;
+  }
+
+  addConnectionName(name) {
+    this._connections[name] = new ViewConnection(name, this);
+    return this._connections[name];
+  }
+
+  nameConnection(connection, name) {
+    var idx = this._unnamedConnections.indexOf(connection);
+    assert(idx >= 0);
+    connection._name = name;
+    this._connections[name] = connection;
+    this._unnamedConnections.splice(idx, 1);
+  }
+
+  isResolved() {
+    return this._unnamedConnections.length == 0;
```

Comment:
according to your comment above, resolved Particle should also have an id, so maybe
return this._id !== undefined && ...?